### ISSUE TITLE
[androidx-migration] google_sign_in and firebase_crashlytics dependency/build errors

### ISSUE THREAD

**Issue by ComplexCarbos** [flutter/flutter#42616]
I am having a heck of time upgrading an app heavy with firebase plugins to AndroidX. To troubleshoot, I created a new hello world baseline:

Run flutter upgrade:

> Running flutter doctor...
> Doctor summary (to see all details, run flutter doctor -v):
> [✓] Flutter (Channel stable, v1.9.1+hotfix.4, on Linux, locale en_US.UTF-8)
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
> [✓] Android Studio (version 3.5)
> [✓] VS Code (version 1.39.1)
> [✓] Connected device (1 available)

Install most recent version of Android studio:

> Android Studio 3.5.1
> Build #AI-191.8026.42.35.5900203, built on September 25, 2019
> JRE: 1.8.0_202-release-1483-b49-5587405 amd64
> JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
> Linux 5.0.0-31-generic

Create default flutter program. Select the following options:

> Use androidx* artifacts
> Include Kotlin support
> Include Swift supprt

- Generate counter app.
- Run it to make sure it works
- In android studio, right click Android app folder. Choose flutter, open Android Module in Android Studio
- Allow update to complete. Note I did not upgrade gradle. I did that try that a few times, but get different errors
- Highlight App, Refactor, Migrate to AndroidX
- Click Do Refactor
- Verify `gradle.properties` has:

> android.enableJetifier=true
> android.useAndroidX=true

- Run App to ensure it executes. It does.
- Add  cloud_firestore: ^0.12.9+5
- Update both gradle files as described [here](https://firebase.google.com/docs/android/setup#add_the_sdk)
- Add google-services.json to the Android app folder
- Update the minSdkVersion (Arbitrarily chose 25)
- Run app to ensure it works. It does.
- Stop program. Flutter clean
- Add firebase_auth: ^0.14.0+5
- Run app to ensure it works. It does.
- Stop program. Flutter clean
- Add google_sign_in: ^4.0.7
- Run app to ensure it works. It doesn't

> Execution failed for task ':app:preDebugBuild'.
> > Android dependency 'androidx.swiperefreshlayout:swiperefreshlayout' has different version for the compile (1.0.0-rc01) and runtime (1.0.0) classpath. You should manually set the same version via DependencyResolution

If I remove auth and firestore, keeping just google sign in, the app will compile and run. Google sign in combined with either auth or firestore fails with one of several dependency errors. I tried to add DependencyResolution settings to the gradle file, but for each one add, I get a new one (`androidx.legacy:legacy-support-core-utils`, `androidx.lifecycle:lifecycle-livedata-core`, etc.)

Pubspec:

```yaml
name: test_firebase
description: A new Flutter application.

version: 1.0.0+1

environment:
  sdk: ">=2.1.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter

  cupertino_icons: ^0.1.2
  cloud_firestore: ^0.12.9+5
  firebase_auth: ^0.14.0+5
  google_sign_in: ^4.0.7

dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:
  uses-material-design: true
```

Project Gradle File:

```gradle
buildscript {
    ext.kotlin_version = '1.2.71'
    repositories {
        google()
        jcenter()
    }

    dependencies {
        classpath 'com.android.tools.build:gradle:3.2.1'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath 'com.google.gms:google-services:4.3.2'
    }
}

allprojects {
    repositories {
        google()
        jcenter()
    }
}

rootProject.buildDir = '../build'
subprojects {
    project.buildDir = "${rootProject.buildDir}/${project.name}"
}
subprojects {
    project.evaluationDependsOn(':app')
}

task clean(type: Delete) {
    delete rootProject.buildDir
}
```

App Gradle File

```gradle
def localProperties = new Properties()
def localPropertiesFile = rootProject.file('local.properties')
if (localPropertiesFile.exists()) {
    localPropertiesFile.withReader('UTF-8') { reader ->
        localProperties.load(reader)
    }
}

def flutterRoot = localProperties.getProperty('flutter.sdk')
if (flutterRoot == null) {
    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
}

def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
if (flutterVersionCode == null) {
    flutterVersionCode = '1'
}

def flutterVersionName = localProperties.getProperty('flutter.versionName')
if (flutterVersionName == null) {
    flutterVersionName = '1.0'
}

apply plugin: 'com.android.application'
apply plugin: 'kotlin-android'
apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"

android {
    compileSdkVersion 28

    sourceSets {
        main.java.srcDirs += 'src/main/kotlin'
    }

    lintOptions {
        disable 'InvalidPackage'
    }

    defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId "com.test_firebase"
        minSdkVersion 25
        targetSdkVersion 28
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            // TODO: Add your own signing config for the release build.
            // Signing with the debug keys for now, so `flutter run --release` works.
            signingConfig signingConfigs.debug
        }
    }
}

flutter {
    source '../..'
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0'
}

apply plugin: 'com.google.gms.google-services'
```

**Comment by ComplexCarbos** [comment]
Another quick observation. I am able to compile and run the flutter demo app with firebase_analytics: ^5.0.2. I am not able to include firebase_crashlytics: ^0.1.0+3. Including  crashlytics results in the same behavior as google sign in.

Just like Google sign in, the app will compile and start up if I include crashlytics only. However, as soon as I add in another plugin (Even firebase_core: ^0.4.0+9), I get the same dependency errors.

Interestingly, if I include firebase_crashlytics and google_sign_in only, I get a build error:

`ERROR: [TAG] Failed to resolve variable '${animal.sniffer.version}'`

But the build does complete and demo app starts up.

In summary, in the base starter flutter demo app, I can build, and execute the app with these:
```
  firebase_core: ^0.4.0+9
  cloud_firestore: ^0.12.9+5
  firebase_auth: ^0.14.0+5
  firebase_analytics: ^5.0.2
```

But it fails with dependency errors if I try to include either of these:
```
  google_sign_in: ^4.0.7
  firebase_crashlytics: ^0.1.0+3
```

**Comment by janmoppel** [comment]
Hello @ComplexCarbos !

Thank you for posting an issue.
Can you please also provide the full error output?

**Comment by ComplexCarbos** [comment]
The errors differ based on the combination of plugins. The one listed below is generated when trying to run the flutter demo app with firebase_core and google_sign_in. When I get a free moment today, I'll push this example to git.

```
rich@rich-GS73VR-7RF:~/AndroidStudioProjects/sibf$ flutter clean
Deleting 'build/'.
Deleting '/home/rich/AndroidStudioProjects/sibf/.dart_tool/'.
Deleting '/home/rich/AndroidStudioProjects/sibf/.android/'.
Deleting '/home/rich/AndroidStudioProjects/sibf/.ios/'.
rich@rich-GS73VR-7RF:~/AndroidStudioProjects/sibf$ flutter build apk --debug
Running "flutter pub get" in sibf...                                0.3s
Initializing gradle...                                              0.6s
Resolving dependencies...                                           7.4s
                                                                        
FAILURE: Build failed with an exception.                                
                                                                        
* What went wrong:                                                      
Execution failed for task ':app:preDebugBuild'.                         
> Android dependency 'androidx.asynclayoutinflater:asynclayoutinflater' has different version for the compile (1.0.0-rc01) and runtime (1.0.0) classpath. You should manually set the same version via DependencyResolution
                                                                        
* Try:                                                                  
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
                                                                        
* Get more help at https://help.gradle.org                              
                                                                        
BUILD FAILED in 16s                                                     
Running Gradle task 'assembleDebug'...                                  
Running Gradle task 'assembleDebug'... Done                        17.4s
*******************************************************************************************
The Gradle failure may have been because of AndroidX incompatibilities in this Flutter app.
See https://goo.gl/CP92wY for more information on the problem and how to fix it.
*******************************************************************************************
Gradle task assembleDebug failed with exit code 1
rich@rich-GS73VR-7RF:~/AndroidStudioProjects/sibf$ 

```

**Comment by ComplexCarbos** [comment]
Hopefully you can reproduce with this:

https://github.com/ComplexCarbos/firebase-androidX-test

**Comment by blasten** [comment]
@ComplexCarbos I'm not sure what is depending on `asynclayoutinflater`. In the project where you get this exception "asynclayoutinflater", would you mind running `./gradlew app:dependencies` and pasting the entire output here?

**Comment by ComplexCarbos** [comment]
[out.txt](https://github.com/flutter/flutter/files/3795927/out.txt)

**Comment by ComplexCarbos** [comment]
No longer an issue for me using recent plugins and up-to-date Visual Studio Code
